tests/go-validation/cluster/cluster.go

```diff
@@ -0,0 +1,212 @@
+		Roles:    roles,
+	}
+}
+
+type Cluster struct {
+	v1.ClusterInterface
+}
+
+type ClusterConfig struct {
+	*apisV1.Cluster
+}
+
+func NewRKE2ClusterConfig(clusterName, namespace, cni, cloudCredentialSecretName, kubernetesVersion string, machinePools []apisV1.RKEMachinePool) *ClusterConfig {
+	typeMeta := metav1.TypeMeta{
+		Kind:       "Cluster",
+		APIVersion: "provisioning.cattle.io/v1",
+	}
+
+	//metav1.ObjectMeta
+	objectMeta := metav1.ObjectMeta{
+		Name:      clusterName,
+		Namespace: namespace,
+	}
+
+	etcd := &rkev1.ETCD{
+		DisableSnapshots:     false,
+		S3:                   nil,
+		SnapshotRetention:    5,
+		SnapshotScheduleCron: "0 */5 * * *",
+	}
+
+	chartValuesMap := rkev1.GenericMap{
+		Data: map[string]interface{}{},
+	}
+
+	machineGlobalConfigMap := rkev1.GenericMap{
+		Data: map[string]interface{}{
+			"cni":     cni,
+			"profile": nil,
+		},
+	}
+
+	localClusterAuthEndpoint := rkev1.LocalClusterAuthEndpoint{
+		CACerts: "",
+		Enabled: false,
+		FQDN:    "",
+	}
+
+	upgradeStrategy := rkev1.ClusterUpgradeStrategy{
+		ControlPlaneConcurrency:  "10%",
+		ControlPlaneDrainOptions: rkev1.DrainOptions{},
+		WorkerConcurrency:        "10%",
+		WorkerDrainOptions:       rkev1.DrainOptions{},
+	}
+
+	rkeSpecCommon := rkev1.RKEClusterSpecCommon{
+		ChartValues:              chartValuesMap,
+		MachineGlobalConfig:      machineGlobalConfigMap,
+		ETCD:                     etcd,
+		LocalClusterAuthEndpoint: localClusterAuthEndpoint,
+		UpgradeStrategy:          upgradeStrategy,
+		MachineSelectorConfig:    []rkev1.RKESystemConfig{},
+	}
+
+	rkeConfig := &apisV1.RKEConfig{
+		RKEClusterSpecCommon: rkeSpecCommon,
+		MachinePools:         machinePools,
+	}
+
+	spec := apisV1.ClusterSpec{
+		CloudCredentialSecretName:            cloudCredentialSecretName,
+		KubernetesVersion:                    kubernetesVersion,
+		DefaultPodSecurityPolicyTemplateName: "",
+
+		RKEConfig: rkeConfig,
+	}
+
+	v1Cluster := &apisV1.Cluster{
+		TypeMeta:   typeMeta,
+		ObjectMeta: objectMeta,
+		Spec:       spec,
+	}
+
+	return &ClusterConfig{
+		v1Cluster,
+	}
+}
+
+func NewCluster(namespace string, client *v1.ProvisioningV1Client) *Cluster {
+	clusters := client.Clusters(namespace)
+	return &Cluster{
+		clusters,
+	}
+}
+
+func (c *Cluster) CreateCluster(clusterConfig *ClusterConfig) (*apisV1.Cluster, error) {
+	ctx := context.Background()
+	v1Cluster, err := c.Create(ctx, clusterConfig.Cluster, metav1.CreateOptions{})
+
+	return v1Cluster, err
+}
+
+func (c *Cluster) GetCluster(clusterName string) (*apisV1.Cluster, error) {
+	ctx := context.Background()
+	v1Cluster, err := c.Get(ctx, clusterName, metav1.GetOptions{})
+	return v1Cluster, err
+}
+
+func (c *Cluster) DeleteCluster(clusterName string) error {
+	ctx := context.Background()
+	err := c.Delete(ctx, clusterName, metav1.DeleteOptions{})
+	return err
+}
+
+func ClusterCleanup(ec2Client *aws.EC2Client, clusters []*Cluster, clusterNames []string, nodes []*aws.EC2Node) error {
+	if ec2Client != nil {
```

Comment:
Shouldn't we remove the Clusters before the Nodes?

Reply:
Yeah that makes more sense. I'll make the change

Reply:
Fixed